### ISSUE TITLE
Recursive tasks can generate profiling events on wrong streams (profile corruption)

### ISSUE THREAD

**Issue by omor1** [ICLDisco/parsec#412]
## Describe the bug
Recursive tasks, such as those used for large tiles in DPLASMA or dense tiles in HiCMA, work by starting a local child taskpool and attaching completion of the parent (asynchronous) task to that of the child taskpool. This is done via `parsec_recursivecall`:
```C
static inline int
parsec_recursivecall( parsec_execution_stream_t    *es,
                      parsec_task_t                *task,
                      parsec_taskpool_t            *tp,
                      parsec_recursive_callback     callback,
                      int nbdesc,
                      ... )
{
    parsec_recursive_cb_data_t *cbdata = NULL;
    /* removed for brevity */
    cbdata = (parsec_recursive_cb_data_t *) malloc(/* size calculation*/);
    cbdata->es       = es;
    cbdata->task     = task;
    cbdata->callback = callback;
    cbdata->nbdesc   = nbdesc;
    /* removed for brevity */
    parsec_taskpool_set_complete_callback( tp, &parsec_recursivecall_callback,
                                           (void *)cbdata );
    parsec_context_add_taskpool( es->virtual_process->parsec_context, tp);
    return -1;
}
```

In short, `parsec_recursivecall` allocates and fills a structure for the recursive callback data—including the execution stream `es` where the parent task executed—sets the completion callback for the child taskpool to `parsec_recursivecall_callback`, and adds the child taskpool to the parent task's context. The completion callback `parsec_recursivecall_callback` is given below:
```C
static inline int parsec_recursivecall_callback(parsec_taskpool_t* tp, void* cb_data)
{
    int i, rc = 0;
    parsec_recursive_cb_data_t* data = (parsec_recursive_cb_data_t*)cb_data;
    /* call user callback *before* we complete and release the task */
    data->callback( tp, data );
    rc = __parsec_complete_execution(data->es, data->task);
    /* removed for brevity */
    return rc;
}
```

This completion callback calls the user's custom completion callback—in DPLASMA and HiCMA this is used in `*potrf` to propagate errors—and then the standard PaRSEC execution completion function `__parsec_complete_execution`. For regular tasks, `__parsec_complete_execution` is called by the runtime after the body of the tasks executes, but for asynchronous tasks it is up to the user to call it—in this case, via the runtime-provided recursive wrapper.

`__parsec_complete_execution` does various tasks around cleaning up task execution, including potentially triggering dependency releases. While doing so, it can generate profiling events. These are written to the profiling stream associated with the execution stream argument passed to `__parsec_complete_execution`.

When `__parsec_complete_execution` is called for a normal task during `__parsec_task_progress` after it completes execution of the task body, the execution stream where it is currently executing on is passed as the argument. Inside `parsec_recursivecall_callback` however, the execution stream passed is `data->es`, which is the execution stream where the _parent_ task originally executed, not the execution stream of wherever this callback is being executed—presumably by the last task in the child taskpool to complete. This is **incorrect** and leads to the profile being corrupted.

### To Reproduce
1. Profile a recursive task
2. Be unlucky and have a corrupted profile
3. ???
4. Profit!

The likelihood of this occurring is higher with many recursive tasks mixed with many short tasks in the parent taskpool as well as with many compute cores.

### Expected behavior
Profiles shouldn't get corrupted by recursive tasks.

I don't think there's currently a way to get the _current_ execution stream from inside the completion callback. The most straightforward solution is to change that, so that PaRSEC passes the execution stream where `on_complete` is called to the callback, but that requires changes to the API, which may not be palatable. In any case, the current behavior is simply **wrong**.

## Additional context
This problem has been frustrating to debug, but I'm fairly certain I've determined the cause.

**Comment by therault** [comment]
TLDR; using the profiling stream stored in the es is incorrect in this scenario.

As a side note, there is a similar scenario in the MADNESS-PaRSEC integration, and TTG will produce more issues similar to this one with its multithreaded comm engine.

I propose we make obsolote the exposed profiling stream interface, and keep this purely in TLS variables.

Pro: this kind of bug will forever disappear, the profiling API will be simplified, less runtime structures will be impacted by the PARSEC_PROF_TRACE macro. 

Con: loading a TLS is slower than loading a parameter (because `libparsec.so` is usually a dynamic library, my understanding is that we need to call `__tls_get_addr()` once per invocation of `parsec_profiling_trace_flags_ts()`); we lose the capacity to create additional logical profiling streams separate from the physical thread they execute on (which might be confusing in traces anyway).

**Comment by bosilca** [comment]
Nice catch @omor1. Please check the pending PR (here and one in DPLASMA). Let me know if they address your issue.

**Comment by omor1** [comment]
Oh, I didn't realize that `parsec_my_execution_stream` exists, that's exactly what we need here!

**Comment by omor1** [comment]
The fix from #413 still has issues even after #414, since the completion callback `parsec_recursivecall_callback` can end up being called on the comm thread, which doesn't set the execution stream in a TLS variable. This can probably be fixed by having it do that, but it might still be safer (and doesn't require a TLS lookup) to pass the execution stream to the completion callback explicitly.

**Comment by bosilca** [comment]
I don't see how in the current version `parsec_recursivecall_callback` can be called from the communication stream. The recursive tasks are local, they do not involve any external communications, so the communication thread should never get involved with them. If you have an example I would like to take a look.

**Comment by omor1** [comment]
> I don't see how in the current version `parsec_recursivecall_callback` can be called from the communication stream. The recursive tasks are local, they do not involve any external communications, so the communication thread should never get involved with them. If you have an example I would like to take a look.

Well, the example is that it _occurs_.

`parsec_recursivecall_callback` is set as the `on_complete` callback. `on_complete` is called from within `parsec_check_complete_cb`. `parsec_check_complete_cb` is called from `parsec_taskpool_update_runtime_nbtask` and from `parsec_context_add_taskpool`, in certain cases. The latter we can eliminate, since taskpools are only added by the user (on the main thread, usually) or in the case of recursive tasks, on a compute thread.

So we're left with callers of `parsec_taskpool_update_runtime_nbtask`. Here, things get more complicated, but the primary callers are `__parsec_complete_execution`—when a task completes, on a compute thread—and `remote_dep_dec_flying_messages`, which can be called in various circumstances, some of which are on the communication thread.

`parsec_taskpool_update_runtime_nbtask` is called in `__parsec_complete_execution` only after the `complete_execution` and `release_task` callbacks are executed. `release_task` in particular can generate remote communications, which can increment reference counters. I think that, maybe, if the communications are faster/slower (not sure which) it can cause the completion callback to be called on the communication thread.
